### PR TITLE
[WIP] Pause the frame scheduler during hot reload

### LINKED ISSUE #26503: Reassemble method called too late
Consider a widget that rebuilds very often (typically because of an animation). 
Then consider editing the sources of that widgets through hot-reload. 

It seems that when doing so, there's a chance for `build` method to be called _before_ `reassemble` was, but with the updated content. 

This is a big issue for the following open source library: https://github.com/rrousselGit/flutter_hooks
As it needs to override the `reassemble` behavior to handle potential changes within `build` method.

But if `build` method is called before `reassemble`, then the library won't be ready for it and the app will enter into an invalid state.

### PR DESCRIPTION
**Pull request by dnfield**
Fixes https://github.com/flutter/flutter/issues/26503

This needs tests - WIP until I can add them.  Posting for review for feedback on whether this approach is sound.

Right now, it's possible to do a hot reload, get new content on the device, and have a `build` called before `reassemble` is called.  This is causing problems for flutter_hooks, and seems like a contract violation for `Element`/`State`.

This patch sends a message to the framework to pause scheduling frames via the AppLifecycle.  Effectively, we're simulating the device sending a suspend and a resume message to the app. 

/cc @rrousselGit

### CODE REVIEW

**Comment by dnfield** [comment]
Sorry about the bad diff - it's because I've added a try/finally block around the whole method 😢

**Comment by goderbauer** [comment]
Reviewing with white-spaces ignored makes this a lot simpler: https://github.com/flutter/flutter/pull/26983/files?utf8=%E2%9C%93&diff=split&w=1

File: packages/flutter/lib/src/widgets/binding.dart [hunk 1]
```diff
@@ -303,6 +303,28 @@ mixin WidgetsBinding on BindingBase, SchedulerBinding, GestureBinding, RendererB
...
 
     assert(() {
+      registerSignalServiceExtension(
+        name: 'lifecycleSuspend',
+        callback: () async {
+          // Let the engine know to stop scheduling frames.
+          await SystemChannels.lifecycle.send('AppLifecycleState.suspending');
+          // Let the framework know we're suspended (the message doesn't get echoed
+          // back from the engine).
+          handleAppLifecycleStateChanged(AppLifecycleState.suspending);
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/binding.dart]: We should probably document on `suspending` and `resumed` that these will also fire during hot reload?

Along the same lines: Does overloading this have any unintended side effects? What if an app developer is listing to these to do something, do we want that to happen on hot reload as well?

**Comment by dnfield** [comment]
From offline discussion:

- Add a test to tooling to make sure we send the messages in the right order (e.g. we send a suspend, reassemble, resume in that order - maybe a test as well to make sure we always end up with a resume if bad things happen in the middle)
- Add a test to the framework to exercise this (a widget that fails if build is called before reassemble, and make sure it doesn't fail when we go to reassemble it using the methods the tooling will be calling over the service protocol)

Also from discussion: Instead of using the suspend message, see if we can get a more specific message just for hot reload/restart (and potentially skip stopping the engine animator if that still works).

**Comment by Hixie** [comment]
This is a clever, but somewhat sketchy solution. We really want each pipe to have a clear owner, and faking messages into our own communications channels is just asking for trouble later.

We can have a dedicated "stop servicing engine frame requests" endpoint in the service protocol, that would be a little better.

But fundamentally even that wouldn't really help (consider a Timer running every millisecond that depends on a data structure that you want to update during the reassemble). We really want a way to say "stop all work, wait for reload, reassemble". We could do that by having an endpoint that we set a breakpoint on, which we call before hot reload, then do the hot reload while the process is paused, then resume, and have that endpoint immediately schedule a microtask that runs reassemble.

cc @rmacnak-google @jonahwilliams @zanderso

File: packages/flutter/lib/src/widgets/binding.dart (hunk 1 above)
Comment (jonahwilliams) [reply packages/flutter/lib/src/widgets/binding.dart]: I don't think this is the right approach, it is totally reasonable to have business logic tied to the app lifecycle state.

**Comment by jonahwilliams** [comment]
If we only care about pausing the frame scheduler to enforce build after reassemble, then I agree with hixie, we can add a dedicated channel for this that works similarly to the current lifecycle channel.

As for the for the Timer problem, I have no idea.

**Comment by goderbauer** [comment]
We are going to close this PR for now since the mechanism needs to be reworked. @dnfield will follow-up with a new PR.

**Pull request closed without merging**